workspace_host/interface.js

```diff
@@ -918,19 +921,39 @@ async function sendGradedFilesArchive(workspace_id, res) {
   const zipName = `${workspace.remote_name}-${timestamp}.zip`;
   const workspaceDir = path.join(config.workspaceHostHomeDirRoot, workspace.remote_name, 'current');
 
+  const gradedFiles = (
+    await fg(workspaceSettings.workspace_graded_files, {
+      cwd: workspaceDir,
+      stats: true,
+    })
+  ).filter((file) => contains(workspaceDir, path.join(workspaceDir, file.path)));
+
+  if (gradedFiles.length > config.workspaceMaxGradedFilesCount) {
+    return res.status(500).send({
+      message: `Cannot submit more than ${config.workspaceMaxGradedFilesCount} files from the workspace.`,
+    });
+  }
+  if (_.sumBy(gradedFiles, (file) => file.stats.size) > config.workspaceMaxGradedFilesSize) {
+    return res.status(500).send({
+      message: `Workspace files exceed limit of ${filesize(config.workspaceMaxGradedFilesSize, {
+        base: 2,
+      })}.`,
+    });
+  }
+
   // Stream the archive back to the client as it's generated.
   res.attachment(zipName).status(200);
   const archive = archiver('zip');
   archive.pipe(res);
 
-  for (const file of workspaceSettings.workspace_graded_files) {
+  for (const file of gradedFiles) {
     try {
-      const filePath = path.join(workspaceDir, file);
+      const filePath = path.join(workspaceDir, file.path);
       await fsPromises.lstat(filePath);
-      archive.file(filePath, { name: file });
-      debug(`Sending ${file}`);
+      archive.file(filePath, { name: file.path });
```

Comment:
Should we be checking the `stat` result for `isFile()` to match the code in `lib/workspace.js`? I think we want to explicitly skip anything that is not a regular file?

Reply:
Actually that is not necessary, either here or in lib/workspace.js. By default `fast-glob` only returns regular files (https://www.npmjs.com/package/fast-glob#onlyfiles). I'll remove the test from lib/workspace.js.

Reply:
Oh, very good.

Reply:
Done in [797286c](https://github.com/PrairieLearn/PrairieLearn/pull/6982/commits/797286c5c8c4a2a443b383d4685654d74bc473b6).